src/cluster_legacy.c

```diff
@@ -5986,6 +6226,245 @@ char *clusterNodeGetShardId(clusterNode *node) {
+        }
+        n = clusterLookupNode(c->argv[4]->ptr, sdslen(c->argv[4]->ptr));
+        if (n == NULL) {
+            addReplyErrorFormat(c,"I don't know about node %s", (char*)c->argv[4]->ptr);
+            return false;
+        }
+        if (nodeIsSlave(n)) {
+            addReplyError(c,"Target node is not a master");
+            return false;
+        }
+    } else if (!strcasecmp(c->argv[3]->ptr,"importing") && c->argc >= 5) {
+        if (server.cluster->slots[s] == myself) {
+            addReplyErrorFormat(c, "I'm already the owner of hash slot %u", s);
+            return false;
+        }
+        n = clusterLookupNode(c->argv[4]->ptr, sdslen(c->argv[4]->ptr));
+        if (n == NULL) {
+            addReplyErrorFormat(c,"I don't know about node %s", (char*)c->argv[4]->ptr);
+            return false;
+        }
+        if (nodeIsSlave(n)) {
+            addReplyError(c,"Target node is not a master");
+            return false;
+        }
+    } else if (!strcasecmp(c->argv[3]->ptr,"stable") && c->argc >= 4) {
+        /* Do nothing */
+    } else if (!strcasecmp(c->argv[3]->ptr,"node") && c->argc >= 5) {
+        /* CLUSTER SETSLOT <SLOT> NODE <NODE ID> */
+        n = clusterLookupNode(c->argv[4]->ptr, sdslen(c->argv[4]->ptr));
+        if (!n) {
+            addReplyErrorFormat(c,"Unknown node %s", (char*)c->argv[4]->ptr);
+            return false;
+        }
+        if (nodeIsSlave(n)) {
+            addReplyError(c,"Target node is not a master");
+            return false;
+        }
+        /* If this hash slot was served by 'myself' before to switch
+         * make sure there are no longer local keys for this hash slot. */
+        if (server.cluster->slots[s] == myself && n != myself) {
+            if (countKeysInSlot(s) != 0) {
+                addReplyErrorFormat(c,
+                    "Can't assign hashslot %d to a different node "
+                    "while I still hold keys for this hash slot.", s);
+                return false;
+            }
+        }
+    } else {
+        addReplyError(c,
+            "Invalid CLUSTER SETSLOT action or number of arguments. Try CLUSTER HELP");
+        return false;
+    }
+
+    *slot = s;
+    *node = n;
+    *timeout_ms = t;
+    return true;
+}
+
+void clusterSetSlotCommand(client *c) {
+    int slot;
+    int timeout_ms;
+    clusterNode *n;
+
+    if (!clusterParseSetSlotCommand(c, &slot, &n, &timeout_ms)) return;
+
+    /* Enhance cluster topology change resilience against primary failures by
+     * replicating SETSLOT before execution.
+     *
+     * Cluster topology changes such slot ownership and migrating states must
+     * be replicated to replicas before applying them to the primary. This
+     * guarantees that after a command is successfully executed, the new state
+     * won't be lost due to a primary node failure. The following example
+     * illustrates how a cluster state can be lost during slot ownership
+     * finalization:
```

Comment:
Many of these comments are perfect material for a doc PR.

`topics/cluster-tutorial.md` and maybe also `topics/cluster-spec.md`. If you start a doc PR, it can be useful also for us to read in parallel when reviewing this PR.

You could even use the good old style "Starting with Valkey 8.0, bla bla bla...".